drivers/serial/serial_xen.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * SPDX-License-Identifier:	GPL-2.0+
+ *
+ * (C) 2018 NXP
+ * (C) 2020 EPAM Systems Inc.
+ */
+#include <common.h>
+#include <cpu_func.h>
+#include <serial.h>
+#include <watchdog.h>
+
+#include <asm/xen/hypercall.h>
+
+DECLARE_GLOBAL_DATA_PTR;
+
+#ifndef CONFIG_DM_SERIAL
+static int xen_serial_init(void)
+{
+	return 0;
+}
+
+/*
+ * N.B. We may be running with d-cache off at the moment, but
+ * according to Xen hypercall ABI (see include/public/arch-arm.h)
+ * all the buffers must reside in memory which is mapped as
+ * Normal Inner Write-Back Inner-Shareable.
+ * So, invalidate the data cache, so Xen sees consistent data.
+ */
+static void xen_serial_putc(const char c)
+{
+	invalidate_dcache_range((unsigned long)&c,
+				(unsigned long)&c + 1);
+	(void)HYPERVISOR_console_io(CONSOLEIO_write, 1, (char *)&c);
+}
+
+static void xen_serial_puts(const char *str)
+{
+	int len = strlen(str);
+
+	invalidate_dcache_range((unsigned long)str,
+				(unsigned long)str + len);
+	(void)HYPERVISOR_console_io(CONSOLEIO_write, len, (char *)str);
+}
+
+static int xen_serial_tstc(void)
+{
+	return 0;
+}
+
+static void xen_serial_setbrg(void)
+{
+}
+
+static struct serial_device xen_serial_drv = {
+	.name	= "xen_serial",
+	.start	= xen_serial_init,
+	.stop	= NULL,
+	.setbrg	= xen_serial_setbrg,
+	/*
+	 * FIXME: Non-priveleged domain should not try reading characters via
```

Comment:
An unprivileged guest cannot use _CONSOLEIO_write()_ on production build (unless you implement your own XSM flask). If you expect U-boot to print important message using this interface, then I would recommend to use a different interface (either the PV console or the pl011).

Reply:
Well, this piece of code is only used if driver model is not enabled and intention was to provide this interface **before** we implement PV console. So, at the end of the day PV console is in place

Reply:
I am not sure to understand this... Is the code in use or not? If it is in use what kind of message do you print on it?

Reply:
It is in use if driver model is disabled which is the case for the very first patch. After we add more meat we enable the driver model and PV console is in charge then.

Reply:
Basically we can drop that code at the end

Reply:
A general of rule is we don't add feature that will be dropped by the end of the series. Although, I don't know what's U-boot stance for such things.

Reply:
We have dropped the serial driver without the driver model